### PR TITLE
Added a linting tool for Cura Printers and Profiles

### PR DESCRIPTION
**Pull request by jellespijker**
printer-linter works on definitions, profiles, and meshes; It has various diagnostic checks. With possible suggestions for fixes. It should also be able to fix certain diagnostic issues, and it can be used to format the files according to code-style.

It can output the diagnostics in a yaml file, which can then be used to comment on PR's with suggestions to the author. Future PR.

The settings for the diagnostics and checks are defined in `.printer-linter` and are very self-explanatory.

```
checks:
    diagnostic-mesh-file-extension: true
    diagnostic-mesh-file-size: true
    diagnostic-definition-redundant-override: true
fixes:
    diagnostic-definition-redundant-override: true
format:
    format-definition-bracket-newline: false
    format-definition-paired-coordinate-array: true
    format-definition-sort-keys: true
    format-definition-indent: 4
    format-profile-space-around-delimiters: true
    format-profile-sort-keys: true
diagnostic-mesh-file-size: 1200000
```

* See https://github.com/Ultimaker/Cura/pull/13857 for an example run over a lot files. We shouldn't merge that PR, but it makes it easier to validate the results of this tool
* See https://github.com/Ultimaker/Cura/pull/13858 for a workflow PR, which automatically formats printer definitions and profiles and creates a co-authored commit with the original author and the GH action bot
* See https://github.com/Ultimaker/Cura/pull/13873 for a workflow PR, which automatically reviews a PR and adds code-suggestion if fixes are available.

### CODE REVIEW

File: printer-linter/src/printerlinter/factory.py [hunk 1]
```diff
@@ -6,15 +7,17 @@
 from .linters.meshes import Meshes
 
 
-def create(file, settings) -> Optional[DiagnosticGenerator]:
+def create(file: Path, settings: dict) -> Optional[DiagnosticGenerator]:
+    """ Returns a DiagnosticGenerator depending on the file format """
     if not file.exists():
         return None
-    if ".inst" in file.suffixes and ".cfg" in file.suffixes:
+    elif ".inst" in file.suffixes and ".cfg" in file.suffixes:
```
Comment (jellespijker) [inline printer-linter/src/printerlinter/factory.py:14]: no `elif` after a `return`

File: printer-linter/src/terminal.py [hunk 2]
```diff
@@ -74,6 +75,18 @@ def formatFile(file: Path, settings) -> None:
             config.write(f, space_around_delimiters=settings["format"].get("format-profile-space-around-delimiters", True))
 
 
+def extract_file_paths(paths: List[Path]) -> List[Path]:
```
Comment (jellespijker) [inline printer-linter/src/terminal.py]: I hate to say this but, not according to UM codestyle convention
```suggestion
def extractFilePaths(paths: List[Path]) -> List[Path]:
```

File: printer-linter/src/printerlinter/diagnostic.py [hunk 3]
```diff
@@ -8,13 +23,12 @@ def __init__(self, file, diagnostic_name, message, level, offset, replacements=N
         self.replacements = replacements
 
     def toDict(self) -> dict:
```
Comment (jellespijker) [inline printer-linter/src/printerlinter/diagnostic.py:25]: incomplete return type
```suggestion
    def toDict(self) -> Dict[str, Any[str, int, List[Replacements]]]:
```

**Review by jellespijker — COMMENTED** [review]
@Joeydelarago See comments

File: printer-linter/src/printerlinter/diagnostic.py (hunk 3 above)
Comment (Joeydelarago) [reply printer-linter/src/printerlinter/diagnostic.py]: Unfortunately Any is not subscriptable like that. I typed it as Dict\[str, Any\]

File: printer-linter/src/printerlinter/formatters/def_json_formatter.py [hunk 4]
```diff
@@ -1,29 +1,84 @@
...
+    "name":     "1",
+    "inherits": "3"
+}
+
+METADATA_SORT_PRIORITY = {
+    "visible": "0",
+    "author": "1",
+    "manufacturer": "2",
+    "file_formats": "3",
+    "platform" : "4"
```
Comment (jellespijker) [inline printer-linter/src/printerlinter/formatters/def_json_formatter.py:23]: If you decide to align the `TOP_LEVEL_SORT_PRIORITY` also apply that alignment in the dict below.
```suggestion
    "visible":      "0",
    "author":       "1",
    "manufacturer": "2",
    "file_formats": "3",
    "platform" :    "4",
```
Also added a trailing comment on the last entry, to ensure that any future diff will not change two lines

File: printer-linter/src/printerlinter/formatters/def_json_formatter.py [hunk 5]
```diff
@@ -1,29 +1,84 @@
...
 
 from .formatter import FileFormatter
 
+
+# Dictionary items with matching keys will be sorted as if they were the value
+# Example: "version" will be sorted as if it was "0"
+TOP_LEVEL_SORT_PRIORITY = {
+    "version":  "0",
+    "name":     "1",
+    "inherits": "3"
```
Comment (jellespijker) [inline printer-linter/src/printerlinter/formatters/def_json_formatter.py:15]:
```suggestion
    "inherits": "3",
```
add a trailing comma, to limit the line changes in future diffs, if sort keys are added

File: printer-linter/src/printerlinter/formatters/def_json_formatter.py [hunk 6]
```diff
@@ -1,29 +1,84 @@
...
+    "visible": "0",
+    "author": "1",
+    "manufacturer": "2",
+    "file_formats": "3",
+    "platform" : "4"
+}
+
+
 class DefJsonFormatter(FileFormatter):
     def format(self, file: Path):
```
Comment (jellespijker) [inline printer-linter/src/printerlinter/formatters/def_json_formatter.py:28]: Shouldn't this be a dunder function `__format__`

**Review by Joeydelarago — APPROVED** [review]

**Pull request merged**